### PR TITLE
Handle big In clause

### PR DESCRIPTION
**Pull request by xiangfu0**
tldr:  This PR introduced the special handling for big IN clause(>=1000 values)

1. Use merge sort for big IN/NOT IN predicate
2. Use shared context to keep sorted IN clause, so one sort per server
3. Adding query option `inPredicateSortThreshold` to allow customize the handling strategy, default value is 1000.

Benchmark and thinking process:
For query predicate with a big value set for IN/NOT_IN clause, the binary search of dictionary is not ideal.
Considering the size of dictionary is N and values in the IN clause is M, then the complexity is O(logN * M). 
Proposed two solutions:
1. Using Bloomfilter, the complexity is O(M) for bloomfilter look up +  O(logN * M) * H (hit rate).
5. Merge sort, when value set is sorted, the complexity is O(N * logN) for sorting and O(M+N) for lookup.

Considering the total number of segments is K, then the complexity for:
1. current solution is O(logN * M *K)
2. BloomFilter: O(M * K) + O(logN * M * K * H) 
6. MergeSort: O(M *logM) + O((N + M) *K)

So 2 will be significantly better when the matching docs are small. 3 is mostly linear.

### CODE REVIEW

**Comment by richardstartin** [comment]
The reasoning makes sense, but, out of curiosity, how do you know it's the binary searches that are the explanatory factor in query latency and not, say, merging large numbers of bitmaps?

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java [hunk 1]
```diff
@@ -297,12 +300,18 @@ public static FilterContext getFilter(FunctionContext filterFunction) {
         for (int i = 1; i < numOperands; i++) {
           values.add(getStringValue(operands.get(i)));
         }
+        if (numOperands > THRESHOLD_OF_LARGE_IN_CLAUSE_VALUES) {
+          Collections.sort(values);
```
Comment (richardstartin) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:304]: I'm not sure this is valid unless the column type is `STRING`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java [hunk 2]
```diff
@@ -139,10 +143,40 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
...
+                }
+              }
+            }
+          } else {
+            int valueIdx = 0;
+            int dictIdx = 0;
+            while (valueIdx < values.size() && dictIdx < dictionary.length()) {
+              String value = values.get(valueIdx);
+              String dictValue = dictionary.getStringValue(dictIdx);
+              int compareResult = value.compareTo(dictValue);
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:162]: Is there a way to make use of the UTF-8 comparison pushdown used in binary searches? It's defined on `ValueReader` - maybe it could be exposed on `Dictionary`? Or `Dictionary` could provide methods to fill a set of dictionary ids given a list of a strings?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java [hunk 3]
```diff
@@ -139,10 +143,40 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
         }
         break;
       case STRING:
-        for (String value : values) {
-          int dictId = dictionary.indexOf(value);
-          if (dictId >= 0) {
-            dictIdSet.add(dictId);
+        if (values.size() > RequestContextUtils.THRESHOLD_OF_LARGE_IN_CLAUSE_VALUES) {
```
Comment (richardstartin) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:146]: I think this should consider the ratio of the size to the dictionary cardinality. What if you have 1001 strings in the in clause but 100000 in the dictionary? Binary search is still going to win here.

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java (hunk 1 above)
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java:304]: Also, take a look at galloping intersection here https://github.com/RoaringBitmap/RoaringBitmap/blob/fd54c0a100629bb578946e2a0bf8b62784878fa8/RoaringBitmap/src/main/java/org/roaringbitmap/Util.java#L1028

**Comment by Jackie-Jiang** [comment]
We already pay the cost of sorting for big IN clause, so maybe we should always use merge sort instead of applying bloom filter. A very important factor not mentioned is linear scan vs random access. I feel merge sort could always be faster because of this. Bloom filter can be applied for small IN clause as well (currently it is already applied during segment pruning, and we can probably move that to the predicate evaluator).
Can you please share some numbers for these 3 approaches?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java (hunk 3 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:146]: Also, take a look at galloping intersection here https://github.com/RoaringBitmap/RoaringBitmap/blob/fd54c0a100629bb578946e2a0bf8b62784878fa8/RoaringBitmap/src/main/java/org/roaringbitmap/Util.java#L1028

**Comment by Jackie-Jiang** [comment]
Merge sort can only be applied to sorted dictionary, so we need to add a condition check (mutable dictionary is not sorted)
Also, if on-heap dictionary is available, we should simply do lookup

**Comment by xiangfu0** [comment]
Put some preliminary results here:

Dictionary Size = 40243

| IN clause size | Binary search | w/ BloomFilter | Sort In Clause | Merge Sort |
|---------------|---------------|---------------|---------------|-------------|
|1000 | 0.702665593 | 0.477092053 | 1.261689 | 0.456446174|
|2000 | 0.976426609 | 0.469555765 | 1.874155 | 2.35581563|
|3000 | 1.395156561 | 0.54921677 | 2.016092 | 2.303482286|
|4000 | 1.880918723 | 0.750761251 | 2.0648 | 2.362960455|
|5000 | 2.761192677 | 1.066657953 | 2.763625 | 2.96542384|
|6000 | 2.8582977489999997 | 1.1276038309999998 | 2.564282 | 2.592623574|
|7000 | 3.742143986 | 1.413617315 | 2.552672 | 2.599197506|
|8000 | 4.295347995 | 1.532301406 | 2.577749 | 2.6646445659999998|
|9000 | 4.453696857 | 1.839579689 | 2.453438 | 2.9550764199999997|
|10000 | 5.223760384 | 2.141856571 | 2.821829 | 3.604565735|
|11000 | 5.859459020999999 | 2.753079806 | 3.826208 | 3.017545197|
|12000 | 5.9234032800000005 | 2.296612749 | 3.378312 | 3.146833064|
|13000 | 6.706917585 | 2.536116155 | 3.630198 | 3.1185865600000002|
|14000 | 8.850008513 | 3.5461908650000002 | 4.859677 | 3.913546161|
|15000 | 8.187981332 | 3.143937341 | 5.251953 | 3.225024074|
|16000 | 8.206689118 | 3.1599710709999997 | 4.73979 | 3.47502625|
|17000 | 8.672926212 | 3.4895446430000003 | 5.149827 | 3.453563212|
|18000 | 9.326708343 | 3.653541142 | 5.452587 | 3.8414371270000003|
|19000 | 9.526910202 | 3.919721084 | 6.442396 | 3.824844376|
|20000 | 10.616695003999999 | 4.158644734 | 6.126283 | 3.8659291280000003|
|21000 | 10.534084935000001 | 4.347735899 | 6.333432 | 3.969913246|
|22000 | 11.031193614 | 4.452593734 | 6.968554 | 3.994151205|
|23000 | 11.293810134 | 4.814520738000001 | 7.332608 | 4.147165715|
|24000 | 11.919929299 | 4.9986157229999995 | 7.446203 | 4.218936294|
|25000 | 12.658507581 | 5.000232526 | 7.558624 | 4.9323873890000005|
|26000 | 13.544178467 | 5.49112032 | 8.372786 | 4.894930091|
|27000 | 13.328060074000001 | 5.939238627 | 8.529412 | 4.883065706|
|28000 | 14.697835813 | 5.68747577 | 9.463493 | 5.320589308|

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java]: Yes, need more benchmark around this number.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java]: Thanks! This code block is better. The perf is about 2x than the previous version.

**Comment by richardstartin** [comment]
> Put some preliminary results here:
> 
> Dictionary Size = 40243 Matched Ids = 92
> 
> Note that the ratio of hit is only about 0.23%.
> 
> IN clause size	Binary search time(ms)	w/ BloomFilter time(ms)	Sort In Clause time(ms)	Merge Sort time(ms)	Merge Sort Optimized time(ms)
> 1000	0.702665593	0.477092053	1.261689	0.456446174	0.464422597
> 2000	0.976426609	0.469555765	1.874155	2.35581563	1.03641343
> 3000	1.395156561	0.54921677	2.016092	2.303482286	1.050509881
> 4000	1.880918723	0.750761251	2.0648	2.362960455	1.015883262
> 5000	2.761192677	1.066657953	2.763625	2.96542384	1.091565855
> 6000	2.8582977489999997	1.1276038309999998	2.564282	2.592623574	1.1344038459999999
> 7000	3.742143986	1.413617315	2.552672	2.599197506	1.161713075
> 8000	4.295347995	1.532301406	2.577749	2.6646445659999998	1.262171061
> 9000	4.453696857	1.839579689	2.453438	2.9550764199999997	1.254242863
> 10000	5.223760384	2.141856571	2.821829	3.604565735	1.525939769
> 11000	5.859459020999999	2.753079806	3.826208	3.017545197	1.475243883
> 12000	5.9234032800000005	2.296612749	3.378312	3.146833064	1.623747765
> 13000	6.706917585	2.536116155	3.630198	3.1185865600000002	1.7264183020000001
> 14000	8.850008513	3.5461908650000002	4.859677	3.913546161	1.67060298
> 15000	8.187981332	3.143937341	5.251953	3.225024074	1.582934511
> 16000	8.206689118	3.1599710709999997	4.73979	3.47502625	1.666822424
> 17000	8.672926212	3.4895446430000003	5.149827	3.453563212	1.886811357
> 18000	9.326708343	3.653541142	5.452587	3.8414371270000003	1.7912497809999999
> 19000	9.526910202	3.919721084	6.442396	3.824844376	1.922495875
> 20000	10.616695003999999	4.158644734	6.126283	3.8659291280000003	1.9555868
> 21000	10.534084935000001	4.347735899	6.333432	3.969913246	2.068954043
> 22000	11.031193614	4.452593734	6.968554	3.994151205	2.090246166
> 23000	11.293810134	4.814520738000001	7.332608	4.147165715	2.133099357
> 24000	11.919929299	4.9986157229999995	7.446203	4.218936294	2.147012917
> 25000	12.658507581	5.000232526	7.558624	4.9323873890000005	2.297837296
> 26000	13.544178467	5.49112032	8.372786	4.894930091	2.5079797800000003
> 27000	13.328060074000001	5.939238627	8.529412	4.883065706	2.467563049
> 28000	14.697835813	5.68747577	9.463493	5.320589308	2.632497346

The bloom filter approach doesn't appear to dominate much of the parameter space, so is it worth the complexity?

**Comment by xiangfu0** [comment]
> > Put some preliminary results here:
> > Dictionary Size = 40243 Matched Ids = 92
> > Note that the ratio of hit is only about 0.23%.
> > IN clause size	Binary search time(ms)	w/ BloomFilter time(ms)	Sort In Clause time(ms)	Merge Sort time(ms)	Merge Sort Optimized time(ms)
> > 1000	0.702665593	0.477092053	1.261689	0.456446174	0.464422597
> > 2000	0.976426609	0.469555765	1.874155	2.35581563	1.03641343
> > 3000	1.395156561	0.54921677	2.016092	2.303482286	1.050509881
> > 4000	1.880918723	0.750761251	2.0648	2.362960455	1.015883262
> > 5000	2.761192677	1.066657953	2.763625	2.96542384	1.091565855
> > 6000	2.8582977489999997	1.1276038309999998	2.564282	2.592623574	1.1344038459999999
> > 7000	3.742143986	1.413617315	2.552672	2.599197506	1.161713075
> > 8000	4.295347995	1.532301406	2.577749	2.6646445659999998	1.262171061
> > 9000	4.453696857	1.839579689	2.453438	2.9550764199999997	1.254242863
> > 10000	5.223760384	2.141856571	2.821829	3.604565735	1.525939769
> > 11000	5.859459020999999	2.753079806	3.826208	3.017545197	1.475243883
> > 12000	5.9234032800000005	2.296612749	3.378312	3.146833064	1.623747765
> > 13000	6.706917585	2.536116155	3.630198	3.1185865600000002	1.7264183020000001
> > 14000	8.850008513	3.5461908650000002	4.859677	3.913546161	1.67060298
> > 15000	8.187981332	3.143937341	5.251953	3.225024074	1.582934511
> > 16000	8.206689118	3.1599710709999997	4.73979	3.47502625	1.666822424
> > 17000	8.672926212	3.4895446430000003	5.149827	3.453563212	1.886811357
> > 18000	9.326708343	3.653541142	5.452587	3.8414371270000003	1.7912497809999999
> > 19000	9.526910202	3.919721084	6.442396	3.824844376	1.922495875
> > 20000	10.616695003999999	4.158644734	6.126283	3.8659291280000003	1.9555868
> > 21000	10.534084935000001	4.347735899	6.333432	3.969913246	2.068954043
> > 22000	11.031193614	4.452593734	6.968554	3.994151205	2.090246166
> > 23000	11.293810134	4.814520738000001	7.332608	4.147165715	2.133099357
> > 24000	11.919929299	4.9986157229999995	7.446203	4.218936294	2.147012917
> > 25000	12.658507581	5.000232526	7.558624	4.9323873890000005	2.297837296
> > 26000	13.544178467	5.49112032	8.372786	4.894930091	2.5079797800000003
> > 27000	13.328060074000001	5.939238627	8.529412	4.883065706	2.467563049
> > 28000	14.697835813	5.68747577	9.463493	5.320589308	2.632497346
> 
> The bloom filter approach doesn't appear to dominate much of the parameter space, so is it worth the complexity?

Agreed, so I guess just mergeSort is good enough.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java [hunk 4]
```diff
@@ -279,4 +281,39 @@ protected byte[] getBytes(int dictId) {
   protected byte[] getBuffer() {
     return new byte[_numBytesPerValue];
   }
+
+  public void getDictIds(List<String> values, IntSet dictIds, int inPredicateSparseThreshold,
+      int inPredicateSortThreshold) {
+    if (length() / values.size() > inPredicateSparseThreshold || values.size() < inPredicateSortThreshold) {
+      for (String value : values) {
```
Comment (saurabhd336) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java:288]: Probably a micro-optimization, but here if `values.size() > inPredicateSparseThreshold `, we know values is sorted, so if we could expose the binary search function to accept a lower bound, we can replace this with something like

```
long lastDictId = 0;
for (String value : values) {
        int dictId = indexOf(llastDictId, value);
        if (dictId >= 0) {
          dictIds.add(dictId);
          lastDictId = dictId;
        }
}
```
i.e., since values is sorted, the search space for a value can be reduced further.

The binary search can accept a new param to specify start idx

```
  protected int binarySearch(int lowIdx, long value) {
    int low = lowIdx;
    int high = _length - 1;
    while (low <= high) {
      int mid = (low + high) >>> 1;
      long midValue = _valueReader.getLong(mid);
      if (midValue < value) {
        low = mid + 1;
      } else if (midValue > value) {
        high = mid - 1;
      } else {
        return mid;
      }
    }
    return -(low + 1);
  }
  ```

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java (hunk 3 above)
Comment (saurabhd336) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java]: I was thinking if we can still benefit from the strings being sorted in this scenario if we use something like https://github.com/apache/pinot/pull/10254#discussion_r1106747320?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java (hunk 4 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java:288]: Infact, if we write another binSearch implementation, which returns the index closest (but not greated) to the value being searched (and another indicator of whether the value actually matched), each of these searches will necessarily reduce the search space for the next value in the sorted list.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 5]
```diff
@@ -703,6 +705,49 @@ private BrokerResponseNative handleRequest(long requestId, String query,
...
+  private void handleBigInClause(String rawTableName, Expression expression, int inPredicateSortThreshold) {
+    if (expression.getType() == ExpressionType.FUNCTION) {
+      Function functionCall = expression.getFunctionCall();
+
+      switch (functionCall.getOperator()) {
+        case "IN":
+        case "NOT_IN":
+          List<Expression> operands = functionCall.getOperands();
+          if (operands.size() > inPredicateSortThreshold && isExpressionStringColumn(operands.get(0), rawTableName)) {
+            Collections.sort(operands.subList(1, operands.size()));
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:730]: Expression compare is quite expensive. Also, it might yield unexpected result when the literal type is different, or by any chance some expression is not literal.

We may move the sort to the server side and use the shared values in `QueryContext` to prevent per-segment computation. This way we can always get the correct data type, and process the values properly

Comment (xiangfu0) [reply pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java]: Removed broker side sort, only do it on server side.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java]: Agreed. We should definitely benchmark this. One more thing to consider is about the random lookup vs sequential read for in-memory or on-disk data access.

Considering the Constant of each operation, it's hard to tell which one will be fast given our current numbers, e.g. N>1000, M > 100k.

Then we need to compare: O(N) * O(1 of sequential read) * lg(M) * O(1 of random read) vs (O(N) + O(M)) * O(1 of sequential read)

It's a bit hard to test until we run it on a large scale data set.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java [hunk 6]
```diff
@@ -157,9 +160,11 @@ public static final class DictionaryBasedNotInPredicateEvaluator extends BaseDic
     int[] _matchingDictIds;
     int[] _nonMatchingDictIds;
 
-    DictionaryBasedNotInPredicateEvaluator(NotInPredicate notInPredicate, Dictionary dictionary, DataType dataType) {
+    DictionaryBasedNotInPredicateEvaluator(NotInPredicate notInPredicate, Dictionary dictionary, DataType dataType,
+        QueryContext queryContext) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java]: Annotate it as `Nullable`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java [hunk 7]
```diff
@@ -157,9 +160,11 @@ public static final class DictionaryBasedNotInPredicateEvaluator extends BaseDic
     int[] _matchingDictIds;
     int[] _nonMatchingDictIds;
 
-    DictionaryBasedNotInPredicateEvaluator(NotInPredicate notInPredicate, Dictionary dictionary, DataType dataType) {
+    DictionaryBasedNotInPredicateEvaluator(NotInPredicate notInPredicate, Dictionary dictionary, DataType dataType,
+        @Nullable QueryContext queryContext) {
       super(notInPredicate);
-      _nonMatchingDictIdSet = PredicateUtils.getDictIdSet(notInPredicate, dictionary, dataType);
+      _nonMatchingDictIdSet = PredicateUtils.getDictIdSet(notInPredicate, dictionary, dataType,
+          queryContext);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java:167]: (nit) Should be one line

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java [hunk 8]
```diff
@@ -139,11 +144,22 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
         }
         break;
       case STRING:
-        for (String value : values) {
-          int dictId = dictionary.indexOf(value);
-          if (dictId >= 0) {
-            dictIdSet.add(dictId);
+        if (queryContext == null || values.size() < Integer.parseInt(queryContext.getQueryOptions()
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:147]: (minor) Suggest changing to `<=` for consistency, we usually treat threshold as exclusive

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java [hunk 9]
```diff
@@ -139,11 +144,22 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
...
+                CommonConstants.Broker.Request.QueryOptionValue.DEFAULT_IN_PREDICATE_SORT_THRESHOLD))) {
+          for (String value : values) {
+            int dictId = dictionary.indexOf(value);
+            if (dictId >= 0) {
+              dictIdSet.add(dictId);
+            }
           }
+        } else {
+          List<String> sortedValues =
+              queryContext.getOrComputeSharedValue(List.class, Equivalence.identity().wrap(inPredicate), v -> {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:158]: (minor)
```suggestion
              queryContext.getOrComputeSharedValue(List.class, Equivalence.identity().wrap(inPredicate), k -> {
```

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java [hunk 10]
```diff
@@ -139,11 +144,22 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
...
+                values.sort(String::compareTo);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java:159]: This can potentially cause problem as `values` is shared across segments, and this method is modifying it. We might want to make a clone of the list and sort on the clone.
(minor) This is equivalent to `values.sort(null)`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java [hunk 11]
```diff
@@ -279,4 +281,34 @@ protected byte[] getBytes(int dictId) {
...
+   * Returns the dictionary id for the given sorted values.
+   * @param sortedValues
+   * @param dictIds
+   */
+  public void getDictIds(List<String> sortedValues, IntSet dictIds) {
+    int valueIdx = 0;
+    int dictIdx = 0;
+    byte[] utf8 = null;
+    boolean needNewUtf8 = true;
+    while (valueIdx < sortedValues.size() && dictIdx < length()) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java:295]: Consider caching the size/length to local variable

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java [hunk 12]
```diff
@@ -279,4 +281,34 @@ protected byte[] getBytes(int dictId) {
   protected byte[] getBuffer() {
     return new byte[_numBytesPerValue];
   }
+
+  /**
+   * Returns the dictionary id for the given sorted values.
+   * @param sortedValues
+   * @param dictIds
+   */
+  public void getDictIds(List<String> sortedValues, IntSet dictIds) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BaseImmutableDictionary.java:290]: (minor) Annotate with `@Override`

**Review by Jackie-Jiang — APPROVED** [review]
